### PR TITLE
VZ-5494.  Allow cert-manager updates

### PR DESCRIPTION
**Pull request by michael-cico**
# Description

Allow Cert-Manager configuration updates
- Relax cert-manager validators to allow updates
- Renew certificates when certificate issuer Common Name does not match the ClusterIssuer Common Name
- Renewal requests achieved by updating certificate status field akin to `cmctl` approach
- Clean up failed certificate requests before renewing
- Add more validation around ACME/LetsEncrypt configurations
- Update custom-CA validation to look for secret
- update/add unit tests

Fixes VZ-5494.

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -929,6 +929,7 @@ k8s.io/klog/v2 v2.0.0/go.mod h1:PBfzABfn139FHAV07az/IF9Wp1bkk3vpT2XSJ76fSDE=
 k8s.io/klog/v2 v2.2.0/go.mod h1:Od+F08eJP+W3HUb4pSrPpgp9DGU4GzlpG/TmITuYh/Y=
 k8s.io/klog/v2 v2.3.0 h1:WmkrnW7fdrm0/DMClc+HIxtftvxVIPAhlVwMQo5yLco=
 k8s.io/klog/v2 v2.3.0/go.mod h1:Od+F08eJP+W3HUb4pSrPpgp9DGU4GzlpG/TmITuYh/Y=
+k8s.io/kube-aggregator v0.19.0 h1:rL4fsftMaqkKjaibArYDaBeqN41CHaJzgRJjUB9IrIg=
```
Comment (pfmackin) [inline go.sum:932]: Looks like inadvertent change

File: pkg/constants/constants.go [hunk 2]
```diff
@@ -47,6 +47,9 @@ const VerrazzanoSystemNamespace = "verrazzano-system"
 // VerrazzanoMultiClusterNamespace is the multi-cluster namespace for Verrazzano
 const VerrazzanoMultiClusterNamespace = "verrazzano-mc"
 
+// CertManagerNamespace - the keycloak namespace
```
Comment (pfmackin) [inline pkg/constants/constants.go:50]: comment says keycloak

File: pkg/k8sutil/k8sutil.go [hunk 3]
```diff
@@ -232,19 +260,27 @@ func ExecPod(client kubernetes.Interface, cfg *rest.Config, pod *v1.Pod, contain
 }
 
 // GetGoClient returns a go-client
-func GetGoClient(log vzlog.VerrazzanoLogger) (kubernetes.Interface, error) {
+func GetGoClient(log ...vzlog.VerrazzanoLogger) (kubernetes.Interface, error) {
+	var logger vzlog.VerrazzanoLogger
+	if len(log) > 0 {
```
Comment (pfmackin) [inline pkg/k8sutil/k8sutil.go:265]: why are we passing multiple loggers?

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 4]
```diff
@@ -135,22 +146,82 @@ func setBashFunc(f bashFuncSig) {
...
+var getClientFunc getCoreV1ClientFuncType = k8sutil.GetCoreV1Client
+
+type getCertManagerClientFuncType func(...string) (certv1client.CertmanagerV1Interface, error)
+
+var getCMClientFunc getCertManagerClientFuncType = k8sutil.GetCertManagerClientset
+
+// renewAllSystemCertificates Update the status field for each certificate generated by the Verrazzano ClusterIssuer
+func renewAllSystemCertificates(cli client.Client, log vzlog.VerrazzanoLogger) error {
+	certList := certv1.CertificateList{}
+	ctx := context.Background()
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:160]: what does Background do?

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 5]
```diff
@@ -135,22 +146,82 @@ func setBashFunc(f bashFuncSig) {
...
+		return err
 	}
-	if !isCAValue {
-		// Create resources needed for Acme certificates
-		err := createOrUpdateAcmeResources(compContext)
-		if err != nil {
-			return compContext.Log().ErrorfNewErr("Failed creating Acme resources: %v", err)
+	for index, currentCert := range certList.Items {
+		if currentCert.Spec.IssuerRef.Name != verrazzanoClusterIssuerName {
+			log.Infof("Certificate %s/%s not issued by the Verrazzano cluster issuer, skipping", currentCert.Namespace, currentCert.Name)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:173]: If this is called in a reconcile loop then use log.Oncef

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 6]
```diff
@@ -135,22 +146,82 @@ func setBashFunc(f bashFuncSig) {
...
+			return err
+		}
+	}
+	return nil
+}
+
+func renewCertificate(ctx context.Context, cmclientv1 certv1client.CertmanagerV1Interface, log vzlog.VerrazzanoLogger, updateCert *certv1.Certificate) error {
+	// Update the certificate status to start a renewal; avoid using controllerruntime.CreateOrUpdate(), while
+	// it should only do an update we don't want to accidentally create a updateCert
+	log.Infof("Updating certificate %s/%s", updateCert.Namespace, updateCert.Name)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:186]: If this is called in a reconcile loop then use log.Oncef

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 7]
```diff
@@ -135,22 +146,82 @@ func setBashFunc(f bashFuncSig) {
...
+
+	// If there are any failed certificate requests, they will block a renewal attempt; delete those
+	if err := cleanupFailedCertificateRequests(ctx, cmclientv1, log, updateCert); err != nil {
+		return err
+	}
+
+	// Set the certificate Issuing condition type to True, per guidance by the CertManager team
+	cmutil.SetCertificateCondition(updateCert, certv1.CertificateConditionIssuing, certmetav1.ConditionTrue,
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:194]: Is this a blocking call?   
If the reconciler is called again will this be called again?  If so, will that be a problem?

File: pkg/k8sutil/k8sutil.go (hunk 3 above)
Comment (michael-cico) [reply pkg/k8sutil/k8sutil.go:265]: we're not.  I wanted the logger to be optional.

File: go.sum (hunk 1 above)
Comment (michael-cico) [reply go.sum:932]: No, it seems to be required in order for the code I added to work. It seems to be the same version of kube-aggregator on the line below.  I had a build error when I tried to build without it.

File: pkg/constants/constants.go (hunk 2 above)
Comment (michael-cico) [reply pkg/constants/constants.go:50]: fixed next push

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go (hunk 4 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:160]: I saw it being used in the app operator, and it is supposed to be not canceled or have any deadline, etc.  However, it seems identical to `context.TODO` if you look at them.  So I'll switch to that.

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go (hunk 5 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:173]: I'm going to switch it to debug, thanks.  I had it as info for debugging initially.

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go (hunk 6 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:186]: OK.

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go (hunk 7 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:194]: No, it's not blocking.  It updates the cert status in the object, and later we call Update to save it in the cluster.

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go [hunk 8]
```diff
@@ -76,8 +75,8 @@ func (c certManagerComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.V
 	if c.IsEnabled(old) && !c.IsEnabled(new) {
 		return fmt.Errorf("Disabling component %s is not allowed", ComponentJSONName)
 	}
-	if !reflect.DeepEqual(c.getCertificateSettings(old), c.getCertificateSettings(new)) {
-		return fmt.Errorf("Updates to certificate settings not allowed for %s", c.GetJSONName())
+	if _, err := validateConfiguration(new); err != nil {
```
Comment (shihchang) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go:78]: certmanager_component should implement/shadow the "func (c certManagerComponent)  ValidateInstall(vz *vzapi.Verrazzano)" to call validateConfiguration(vz) as well

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 9]
```diff
@@ -135,22 +151,121 @@ func setBashFunc(f bashFuncSig) {
...
+	if err != nil {
+		return err
+	}
+	issuerCNs, err := findIssuerCommonName(compContext.EffectiveCR().Spec.Components.CertManager.Certificate, isCAConfig)
+	if err != nil {
+		return err
+	}
+	for index, currentCert := range certList.Items {
+		if currentCert.Spec.IssuerRef.Name != verrazzanoClusterIssuerName {
+			log.Debugw("Certificate %s/%s not issued by the Verrazzano cluster issuer, skipping", currentCert.Namespace, currentCert.Name)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:198]: Better to use log.Oncef than Debug if it is somewhat important to see.

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:203]: OK

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go (hunk 8 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go:78]: fixed next pass

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 10]
```diff
@@ -482,8 +689,57 @@ func createOrUpdateCAResources(compContext spi.ComponentContext) error {
...
+		block, rest := pem.Decode(certBytes)
+		if block == nil {
+			break
+		}
+		if block.Type == "CERTIFICATE" {
+			// we're only interested in the leaf cert for the CA, as it is the issuing authority in the cluster
+			leafCACertBytes = block.Bytes
+			break
+		}
+		certBytes = rest
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:738]: if `pem.Decode` never returns `nil` for `block`, which shouldn't happen.

**Pull request merged**